# HTIN5005 Assignment 1 — Part B: PLM-ICD Reimplementation
## Julia Windegger — Student ID: 560823325

This notebook reimplements the **PLM-ICD** baseline (Huang, Tsai & Chen, 2022) as the core-experiment reproduction for the reviewed paper *"Automated ICD Coding using Extreme Multilabel Long Text Transformer-based Models"* (Liu et al.), which uses PLM-ICD as its baseline before building XR-LAT on top of it.

**Source code attribution:** Training script adapted from `MiuLab/PLM-ICD` (Huang, Chao-Wei, Shang-Chi Tsai, and Yun-Nung Chen. 2022. *PLM-ICD: Automatic ICD Coding with Pretrained Language Models*. Proceedings of the 4th Clinical NLP Workshop, ACL. https://github.com/MiuLab/PLM-ICD).

**What this notebook does:** It downloads the original authors' code for automatically assigning diagnosis codes to hospital discharge notes, and reimplements two of their key experimental configurations. Because the underlying model can only process a limited amount of text at once, each discharge note is split into fixed-size chunks before being read — so the two configurations differ not just in how much of the note is read overall, but in how that text is chunked. One configuration reads a shorter input of 3,072 tokens (a token being roughly one word or word-fragment) split into small 128-token chunks, and the other reads a substantially longer input of 5,120 tokens split into larger 512-token chunks. This lets the notebook test whether reading more of the note, and reading it in larger chunks, improves the model's predictions.

**Data note:** Real MIMIC-III hospital note data requires a formal, credentialed application (PhysioNet + CITI ethics training) that can take time to be approved. Since approval was not granted in time, this notebook built a synthetic (but carefully designed to fully understand the model and its setup) dataset with the required file format, so the parts of the pipeline can be built, tested and demonstrated.

**AI Use Disclaimer:**
Parts of the code in this notebook were developed with the assistance of an AI tool, used in an iterative, conversational manner throughout the project. Rather than generating the code independently, I directed each step: I described the specific goal I was trying to reach (such as adapting the PLM-ICD training script, generating a synthetic dataset, or extracting evaluation metrics), and the AI acted as a coding assistant, suggesting implementations, explaining how existing code worked, and helping debug errors as they arose. All resulting code was reviewed, tested, and adapted by me to fit the requirements of this assignment. This assistance also supported my own understanding of the base paper's model architecture and implementation along the way.

##  1. Set up the environment
Before any code can run, we need: a persistent storage location (Google Drive), the original researchers' code, and the right software libraries installed. These three steps only need to be understood once — they don't relate to the ICD coding task itself, they're just getting the workshop ready.

### 1.1 Connect to Google Drive and download the researchers' code

Google Colab's temporary storage is wiped every time the session disconnects. Connecting to Google Drive first means the downloaded code, and anything we build on top of it, will still be there next time.

The check below (`if not os.path.exists(...)`) means this only downloads the code once — if it is already there from an earlier session, it skips straight past this step instead of overwriting it.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import os
import shutil
import tempfile

REPO_DIR = "/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW"

if not os.path.exists(os.path.join(REPO_DIR, "src", "run_icd.py")):
    os.makedirs(REPO_DIR, exist_ok=True)
    with tempfile.TemporaryDirectory() as tmp_dir:
        clone_path = os.path.join(tmp_dir, "PLM-ICD")
        !git clone https://github.com/MiuLab/PLM-ICD.git "{clone_path}"
        for item in os.listdir(clone_path):
            src = os.path.join(clone_path, item)
            dst = os.path.join(REPO_DIR, item)
            if os.path.exists(dst):
                if os.path.isdir(dst):
                    shutil.rmtree(dst)
                else:
                    os.remove(dst)
            shutil.move(src, dst)
    print("Cloned into Drive (merged into existing folder).")
else:
    print("Repo already exists in Drive — skipping clone so your edits aren't overwritten.")

%cd {REPO_DIR}

Repo already exists in Drive — skipping clone so your edits aren't overwritten.
/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW


### 1.2 Install the required software libraries

The researchers' code was written in 2021 and lists very old library versions that no longer install on today's Colab. Instead, we install current, working versions here.

One specific choice worth explaining: `transformers` (the library that provides the pretrained language model) is deliberately locked to version 4.46.3, not the newest release. A newer version changed something internally that breaks this particular code's custom model, so this version is the newest one confirmed to still work with it.

This step needs to be repeated every time you start a fresh Colab session — installed software doesn't get saved to Drive, only files do.

In [3]:
!pip install -q torch transformers==4.46.3 datasets accelerate scikit-learn pandas numpy scipy tqdm

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 44.1/44.1 kB 2.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.0/10.0 MB 102.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 566.4/566.4 kB 44.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 74.2 MB/s eta 0:00:00
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
gradio 6.27.0 requires huggingface-hub<2.0,>=1.16.0, but you have huggingface-hub 0.36.2 which is incompatible.
diffusers 0.40.0 requires huggingface-hub<2.0,>=1.23.0, but you have huggingface-hub 0.36.2 which is incompatible.


### 1.3 Confirm a GPU is available

Training even a small version of this model on a regular processor (CPU) would be far too slow to be practical. This checks that Colab has assigned a GPU (graphics card) to this session before continuing.

If this prints "CPU only", go to Runtime > Change runtime type > select GPU, then come back and re-run this cell.

In [4]:
import torch
print("CUDA available:", torch.cuda.is_available())
print("Device:", torch.cuda.get_device_name(0) if torch.cuda.is_available() else "CPU only")

CUDA available: True
Device: Tesla T4


## 2. Build a synthetic dataset (stand-in for real hospital notes)

Real MIMIC-III data requires official approval that can take days to weeks to come through, so this step creates a synthetic but carefully designed dataset in the meantime. It has the same file format the model expects, so the exact same training code will work later once real data is available - nothing downstream needs to change.

Two things make this fake dataset actually useful for testing (rather than just a placeholder):

1. **Each diagnosis code is tied to a specific sentence.** For example, the code for high blood pressure is always paired with a sentence describing elevated blood pressure readings. This gives the model something real to learn, instead of trying to predict labels that have no connection to the text at all.
2. **That sentence can appear anywhere in a long note** - sometimes near the top, sometimes buried deep in a long block of unrelated filler text. This matters because the whole point of the next section is to compare a model that only reads a *short* piece of each note against one that reads a *much longer* piece. If every important sentence were near the top, both models would do equally well and the comparison would tell us nothing. By hiding some sentences deep in the note, only the model that reads further should be able to find them.

Because the text and labels here are artificially constructed rather than real clinical writing, any accuracy numbers produced from this data are a **demonstration that the pipeline and comparison work correctly** - not a measurement of real diagnostic performance. That will only be meaningful once trained on real MIMIC-III notes.

In [ ]:
import os
import random
import pandas as pd

DATA_DIR = f"{REPO_DIR}/data/mimic3_v1"

if not os.path.exists(os.path.join(DATA_DIR, "train_50.csv")):

    random.seed(42)

    ALL_CODES_50 = [
        "401.9", "428.0", "427.31", "414.01", "584.9", "250.00", "486", "285.9",
        "38.93", "V58.61", "038.9", "996.62", "599.0", "244.9", "530.81",
        "272.4", "276.1", "285.1", "V45.81", "36.15", "99.04", "88.72",
        "V15.82", "574.20", "998.59", "39.61", "96.71", "36.06",
        "410.71", "780.09", "997.1", "424.0", "V45.82", "45.13", "584.5",
        "45.16", "276.2", "424.1", "511.9", "412", "N39.0", "78.52",
        "790.29", "780.2", "008.45", "790.6", "263.9", "V44.21", "560.1",
    ]
    ALL_CODES_50 = list(dict.fromkeys(ALL_CODES_50))[:50]

    CODE_SIGNALS = {
        "401.9":  "Blood pressure remains persistently elevated at 165/95 on repeat measurement.",
        "428.0":  "Echocardiogram demonstrates reduced ejection fraction consistent with heart failure.",
        "427.31": "Telemetry reveals an irregularly irregular rhythm consistent with atrial fibrillation.",
        "414.01": "Coronary angiogram shows significant stenosis of the left anterior descending artery.",
        "584.9":  "Creatinine has risen sharply from baseline, consistent with acute kidney injury.",
        "250.00": "Fingerstick glucose readings remain uncontrolled despite standing insulin regimen.",
        "486":    "Chest radiograph demonstrates a new consolidation consistent with pneumonia.",
        "285.9":  "Hemoglobin is low without an identified source of active bleeding.",
        "38.93":  "A central venous catheter was placed under ultrasound guidance without complication.",
        "V58.61": "Patient continues on long-term anticoagulation for prior thromboembolic event.",
        "038.9":  "Blood cultures returned positive, consistent with bloodstream infection.",
        "996.62": "Surgical site shows erythema and purulent drainage concerning for infected hardware.",
        "599.0":  "Urinalysis demonstrates pyuria and bacteriuria consistent with urinary tract infection.",
        "244.9":  "TSH is markedly elevated, consistent with underlying hypothyroidism.",
        "530.81": "Patient reports worsening reflux symptoms with regurgitation after meals.",
        "272.4":  "Lipid panel shows markedly elevated LDL cholesterol.",
        "276.1":  "Serum sodium returned low at 128, consistent with hyponatremia.",
        "285.1":  "Acute drop in hemoglobin following procedure, consistent with acute blood loss anemia.",
        "V45.81": "Patient has a prior coronary artery bypass graft noted on surgical history.",
        "36.15":  "Patient underwent single-vessel coronary artery bypass grafting during this admission.",
        "99.04":  "Patient received a packed red blood cell transfusion for symptomatic anemia.",
        "88.72":  "Transthoracic echocardiogram was performed to assess cardiac function.",
        "V15.82": "Documented history of tobacco use is noted in the social history.",
        "574.20": "Right upper quadrant ultrasound demonstrates cholelithiasis without wall thickening.",
        "998.59": "Post-operative course complicated by a wound infection requiring antibiotics.",
        "39.61":  "Patient required cardiopulmonary bypass during the operative procedure.",
        "96.71":  "Patient was placed on mechanical ventilation for respiratory failure.",
        "36.06":  "A drug-eluting stent was placed in the culprit coronary vessel.",
        "410.71": "Troponin is elevated with dynamic ST-segment changes, consistent with myocardial infarction.",
        "780.09": "Patient exhibited an altered level of consciousness on initial evaluation.",
        "997.1":  "Post-procedural course was complicated by a cardiac complication requiring monitoring.",
        "424.0":  "Echocardiogram demonstrates severe mitral regurgitation.",
        "V45.82": "History notes prior percutaneous coronary intervention with stent placement.",
        "45.13":  "Upper endoscopy was performed to evaluate the source of gastrointestinal symptoms.",
        "584.5":  "Acute kidney injury is attributed to reduced renal perfusion in the setting of illness.",
        "45.16":  "Colonoscopy was performed as part of the diagnostic workup.",
        "276.2":  "Arterial blood gas demonstrates metabolic acidosis with elevated lactate.",
        "424.1":  "Auscultation reveals a murmur consistent with aortic valve pathology.",
        "511.9":  "Chest imaging demonstrates a pleural effusion of uncertain etiology.",
        "412":    "History is notable for a prior myocardial infarction with no current acute changes.",
        "N39.0":  "Repeat urinalysis remains positive, consistent with a recurrent urinary tract infection.",
        "78.52":  "Patient required emergent cardiopulmonary resuscitation during this admission.",
        "790.29": "Point-of-care glucose testing reveals abnormal readings requiring further workup.",
        "780.2":  "Patient experienced a syncopal episode prior to admission.",
        "008.45": "Stool studies are positive for Clostridioides difficile infection.",
        "790.6":  "Basic metabolic panel reveals an electrolyte abnormality requiring correction.",
        "263.9":  "Nutritional assessment indicates evidence of malnutrition.",
        "V44.21": "Patient care requires management of a tracheostomy present from prior admission.",
        "560.1":  "Abdominal imaging is consistent with a bowel obstruction.",
    }

    FILLER_SENTENCES = [
        "Patient remains hemodynamically stable overnight without acute events.",
        "Vital signs are within normal limits at this time of assessment.",
        "Nursing notes indicate the patient tolerated the shift without incident.",
        "Physical therapy evaluated the patient and recommended continued mobility exercises.",
        "Patient reports adequate pain control on the current medication regimen.",
        "Family was updated at bedside regarding the current plan of care.",
        "Intake and output remain balanced over the past 24-hour period.",
        "Patient was seen and examined; no new complaints at this time.",
        "Diet was advanced as tolerated without difficulty.",
        "Social work is assisting with discharge planning and placement.",
        "Wound care team recommends continued daily dressing changes.",
        "Case management continues to coordinate follow-up appointments.",
        "Patient ambulated in the hallway with standby assistance today.",
        "Overnight course was unremarkable with no acute changes noted.",
        "Medication reconciliation was completed on admission.",
        "Patient and family verbalize understanding of the discharge instructions.",
        "Respiratory status remains stable on room air.",
        "Patient denies any new symptoms since the last assessment.",
        "Sleep was reportedly adequate overnight per nursing documentation.",
        "Continue current plan of care and reassess in the morning.",
    ]

    def build_note(assigned_codes):
        segments = []
        segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(2, 6)))

        signals_to_place = [CODE_SIGNALS[code] for code in assigned_codes]
        random.shuffle(signals_to_place)
        for signal_sentence in signals_to_place:
            segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(10, 60)))
            segments.append(signal_sentence)

        segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(2, 10)))
        return " ".join(segments)

    CODE_WEIGHTS = [1.0 / (rank + 1) for rank in range(len(ALL_CODES_50))]

    def pick_labels_for_one_note():
        target_count = random.randint(1, 6)
        picked = set()
        while len(picked) < target_count:
            picked.add(random.choices(ALL_CODES_50, weights=CODE_WEIGHTS, k=1)[0])
        return list(picked)

    def make_one_row(admission_id):
        labels = pick_labels_for_one_note()
        return {
            "SUBJECT_ID": random.randint(10000, 99999),
            "HADM_ID": admission_id,
            "text": build_note(labels),
            "label": ";".join(labels),
        }

    def make_split(number_of_notes, starting_id):
        rows = [make_one_row(starting_id + i) for i in range(number_of_notes)]
        return pd.DataFrame(rows, columns=["SUBJECT_ID", "HADM_ID", "text", "label"])

    os.makedirs(DATA_DIR, exist_ok=True)

    train_df = make_split(300, starting_id=1)
    dev_df   = make_split(60,  starting_id=1000)
    test_df  = make_split(60,  starting_id=2000)

    train_df.to_csv(os.path.join(DATA_DIR, "train_50.csv"), index=False)
    dev_df.to_csv(os.path.join(DATA_DIR, "dev_50.csv"), index=False)
    test_df.to_csv(os.path.join(DATA_DIR, "test_50.csv"), index=False)

    with open(os.path.join(DATA_DIR, "ALL_CODES_50.txt"), "w") as f:
        f.write("\n".join(ALL_CODES_50))

    print("Dataset created:")
    for fname in ["train_50.csv", "dev_50.csv", "test_50.csv", "ALL_CODES_50.txt"]:
        print(" -", os.path.join(DATA_DIR, fname))
    print("\nExample note length (words):", len(train_df.iloc[0]["text"].split()))
    print("Example labels:", train_df.iloc[0]["label"])

else:
    print("Dataset already exists - skipping regeneration.")
    print("Location:", DATA_DIR)

Dataset created:
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v1/train_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v1/dev_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v1/test_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v1/ALL_CODES_50.txt

Example note length (words): 1785
Example labels: 285.9;530.81;272.4;401.9;428.0;250.00


### 2.1 Run One: train with a short reading window (BL-1 setting)

This reproduces the original PLM-ICD paper's setting: the model reads the first 3,072 tokens of each note, split into twenty-four 128-token pieces.

Since some of our signal sentences are deliberately placed further into the note than that, this model is expected to miss some of them - which is the whole point of the comparison in the next section.

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v1/train_50.csv \
    --validation_file ../data/mimic3_v1/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v1/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 1 \
    --num_warmup_steps 10 \
    --output_dir ../models/mock-test-v1-bl1 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 01:51:02.341594: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Generating train split: 300 examples [00:00, 6367.23 examples/s]
Generating validation split: 60 examples [00:00, 4751.14 examples/s]
config.json: 100% 385/385 [00:00<00:00, 3.34MB/s]
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,


### 2.2 Run Two: train with a much longer reading window (BL-5 setting)

Everything is identical to Run 1 above, except the model now reads up to 5,120 tokens of each note, split into ten 512-token pieces instead of twenty-four 128-token pieces. This is the paper's best-performing configuration.

Because chunk length and reading window are the only settings changed between Run 1 and Run 2, any difference in the results can be attributed specifically to how much of the note the model was able to see,  which is exactly what this comparison is meant to test.

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v1/train_50.csv \
    --validation_file ../data/mimic3_v1/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v1/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 1 \
    --num_warmup_steps 10 \
    --output_dir ../models/mock-test-v1-bl5 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 01:55:11.180239: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 2.3 First check: comparing the outcome for BL-1 vs. BL-5

BL-1 (3,072-token window) and BL-5 (5,120-token window) produce essentially identical results on the v1 dataset. Macro and micro F1 match exactly (0.015 and 0.257), and the remaining metrics differ only marginally with no consistent direction in BL-5's favour.

This isn't a pipeline bug, it means the two models were reading largely the same content. There are two plausible explanations for that: either the notes generated so far are too short for the difference in reading window to matter, or a single training epoch isn't enough for the models to develop distinguishable confidence in their predictions.

Dataset length is the simpler factor to rule out first, so the next section increases the amount of filler text placed between signal sentences, so notes are reliably long enough for some signals to fall beyond BL-1's 3,072-token window. Whether training length also plays a role will be tested afterwards.

In [ ]:
import subprocess
import re
import ast
import pandas as pd

def get_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Re-runs evaluation only (num_train_epochs 0) using the already-trained
    # model saved at model_dir, instead of training again. This reproduces
    # the same metrics that were printed during training, captured directly
    # from the command's output this time instead of being retyped by hand.
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/dev_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]

    result = subprocess.run(
        command,
        cwd=f"{REPO_DIR}/src",
        capture_output=True,
        text=True,
    )

    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")

    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

bl1_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v1-bl1",
    data_dir=f"{REPO_DIR}/data/mimic3_v1",
    max_length=3072,
    chunk_size=128,
)

bl5_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v1-bl5",
    data_dir=f"{REPO_DIR}/data/mimic3_v1",
    max_length=5120,
    chunk_size=512,
)

comparison_df = pd.DataFrame(
    {"BL-1 (3,072 tokens)": bl1_metrics, "BL-5 (5,120 tokens)": bl5_metrics}
).T
comparison_df = comparison_df[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
comparison_df

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1 (3,072 tokens)",0.015306,0.257143,0.539581,0.769484,0.594167,0.25625,0.358072
"BL-5 (5,120 tokens)",0.015306,0.257143,0.526775,0.781198,0.582778,0.26250,0.361962


## 3. Build adjusted dataset — longer padding before each signal sentence

In [ ]:
import os
import random
import pandas as pd

DATA_DIR = f"{REPO_DIR}/data/mimic3_v2"

if not os.path.exists(os.path.join(DATA_DIR, "train_50.csv")):

    random.seed(42)

    ALL_CODES_50 = [
        "401.9", "428.0", "427.31", "414.01", "584.9", "250.00", "486", "285.9",
        "38.93", "V58.61", "038.9", "996.62", "599.0", "244.9", "530.81",
        "272.4", "276.1", "285.1", "V45.81", "36.15", "99.04", "88.72",
        "V15.82", "574.20", "998.59", "39.61", "96.71", "36.06",
        "410.71", "780.09", "997.1", "424.0", "V45.82", "45.13", "584.5",
        "45.16", "276.2", "424.1", "511.9", "412", "N39.0", "78.52",
        "790.29", "780.2", "008.45", "790.6", "263.9", "V44.21", "560.1",
    ]
    ALL_CODES_50 = list(dict.fromkeys(ALL_CODES_50))[:50]

    CODE_SIGNALS = {
        "401.9":  "Blood pressure remains persistently elevated at 165/95 on repeat measurement.",
        "428.0":  "Echocardiogram demonstrates reduced ejection fraction consistent with heart failure.",
        "427.31": "Telemetry reveals an irregularly irregular rhythm consistent with atrial fibrillation.",
        "414.01": "Coronary angiogram shows significant stenosis of the left anterior descending artery.",
        "584.9":  "Creatinine has risen sharply from baseline, consistent with acute kidney injury.",
        "250.00": "Fingerstick glucose readings remain uncontrolled despite standing insulin regimen.",
        "486":    "Chest radiograph demonstrates a new consolidation consistent with pneumonia.",
        "285.9":  "Hemoglobin is low without an identified source of active bleeding.",
        "38.93":  "A central venous catheter was placed under ultrasound guidance without complication.",
        "V58.61": "Patient continues on long-term anticoagulation for prior thromboembolic event.",
        "038.9":  "Blood cultures returned positive, consistent with bloodstream infection.",
        "996.62": "Surgical site shows erythema and purulent drainage concerning for infected hardware.",
        "599.0":  "Urinalysis demonstrates pyuria and bacteriuria consistent with urinary tract infection.",
        "244.9":  "TSH is markedly elevated, consistent with underlying hypothyroidism.",
        "530.81": "Patient reports worsening reflux symptoms with regurgitation after meals.",
        "272.4":  "Lipid panel shows markedly elevated LDL cholesterol.",
        "276.1":  "Serum sodium returned low at 128, consistent with hyponatremia.",
        "285.1":  "Acute drop in hemoglobin following procedure, consistent with acute blood loss anemia.",
        "V45.81": "Patient has a prior coronary artery bypass graft noted on surgical history.",
        "36.15":  "Patient underwent single-vessel coronary artery bypass grafting during this admission.",
        "99.04":  "Patient received a packed red blood cell transfusion for symptomatic anemia.",
        "88.72":  "Transthoracic echocardiogram was performed to assess cardiac function.",
        "V15.82": "Documented history of tobacco use is noted in the social history.",
        "574.20": "Right upper quadrant ultrasound demonstrates cholelithiasis without wall thickening.",
        "998.59": "Post-operative course complicated by a wound infection requiring antibiotics.",
        "39.61":  "Patient required cardiopulmonary bypass during the operative procedure.",
        "96.71":  "Patient was placed on mechanical ventilation for respiratory failure.",
        "36.06":  "A drug-eluting stent was placed in the culprit coronary vessel.",
        "410.71": "Troponin is elevated with dynamic ST-segment changes, consistent with myocardial infarction.",
        "780.09": "Patient exhibited an altered level of consciousness on initial evaluation.",
        "997.1":  "Post-procedural course was complicated by a cardiac complication requiring monitoring.",
        "424.0":  "Echocardiogram demonstrates severe mitral regurgitation.",
        "V45.82": "History notes prior percutaneous coronary intervention with stent placement.",
        "45.13":  "Upper endoscopy was performed to evaluate the source of gastrointestinal symptoms.",
        "584.5":  "Acute kidney injury is attributed to reduced renal perfusion in the setting of illness.",
        "45.16":  "Colonoscopy was performed as part of the diagnostic workup.",
        "276.2":  "Arterial blood gas demonstrates metabolic acidosis with elevated lactate.",
        "424.1":  "Auscultation reveals a murmur consistent with aortic valve pathology.",
        "511.9":  "Chest imaging demonstrates a pleural effusion of uncertain etiology.",
        "412":    "History is notable for a prior myocardial infarction with no current acute changes.",
        "N39.0":  "Repeat urinalysis remains positive, consistent with a recurrent urinary tract infection.",
        "78.52":  "Patient required emergent cardiopulmonary resuscitation during this admission.",
        "790.29": "Point-of-care glucose testing reveals abnormal readings requiring further workup.",
        "780.2":  "Patient experienced a syncopal episode prior to admission.",
        "008.45": "Stool studies are positive for Clostridioides difficile infection.",
        "790.6":  "Basic metabolic panel reveals an electrolyte abnormality requiring correction.",
        "263.9":  "Nutritional assessment indicates evidence of malnutrition.",
        "V44.21": "Patient care requires management of a tracheostomy present from prior admission.",
        "560.1":  "Abdominal imaging is consistent with a bowel obstruction.",
    }

    FILLER_SENTENCES = [
        "Patient remains hemodynamically stable overnight without acute events.",
        "Vital signs are within normal limits at this time of assessment.",
        "Nursing notes indicate the patient tolerated the shift without incident.",
        "Physical therapy evaluated the patient and recommended continued mobility exercises.",
        "Patient reports adequate pain control on the current medication regimen.",
        "Family was updated at bedside regarding the current plan of care.",
        "Intake and output remain balanced over the past 24-hour period.",
        "Patient was seen and examined; no new complaints at this time.",
        "Diet was advanced as tolerated without difficulty.",
        "Social work is assisting with discharge planning and placement.",
        "Wound care team recommends continued daily dressing changes.",
        "Case management continues to coordinate follow-up appointments.",
        "Patient ambulated in the hallway with standby assistance today.",
        "Overnight course was unremarkable with no acute changes noted.",
        "Medication reconciliation was completed on admission.",
        "Patient and family verbalize understanding of the discharge instructions.",
        "Respiratory status remains stable on room air.",
        "Patient denies any new symptoms since the last assessment.",
        "Sleep was reportedly adequate overnight per nursing documentation.",
        "Continue current plan of care and reassess in the morning.",
    ]

    def build_note(assigned_codes):
        segments = []
        segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(2, 6)))

        signals_to_place = [CODE_SIGNALS[code] for code in assigned_codes]
        random.shuffle(signals_to_place)
        for signal_sentence in signals_to_place:
            # Increased from randint(10, 60) to randint(30, 80): the v1 range
            # only pushed a small fraction of notes past the 3,072-token
            # cutoff by chance. Raising the minimum to 30 (and the max to 80)
            # makes notes consistently longer, so BL-1 should genuinely miss
            # more signal sentences that BL-5 can still reach - while staying
            # comfortably clear of BL-5's own 5,120-token limit.
            segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(30, 80)))
            segments.append(signal_sentence)

        segments.extend(random.choices(FILLER_SENTENCES, k=random.randint(2, 10)))
        return " ".join(segments)

    CODE_WEIGHTS = [1.0 / (rank + 1) for rank in range(len(ALL_CODES_50))]

    def pick_labels_for_one_note():
        target_count = random.randint(1, 6)
        picked = set()
        while len(picked) < target_count:
            picked.add(random.choices(ALL_CODES_50, weights=CODE_WEIGHTS, k=1)[0])
        return list(picked)

    def make_one_row(admission_id):
        labels = pick_labels_for_one_note()
        return {
            "SUBJECT_ID": random.randint(10000, 99999),
            "HADM_ID": admission_id,
            "text": build_note(labels),
            "label": ";".join(labels),
        }

    def make_split(number_of_notes, starting_id):
        rows = [make_one_row(starting_id + i) for i in range(number_of_notes)]
        return pd.DataFrame(rows, columns=["SUBJECT_ID", "HADM_ID", "text", "label"])

    os.makedirs(DATA_DIR, exist_ok=True)

    train_df = make_split(300, starting_id=1)
    dev_df   = make_split(60,  starting_id=1000)
    test_df  = make_split(60,  starting_id=2000)

    train_df.to_csv(os.path.join(DATA_DIR, "train_50.csv"), index=False)
    dev_df.to_csv(os.path.join(DATA_DIR, "dev_50.csv"), index=False)
    test_df.to_csv(os.path.join(DATA_DIR, "test_50.csv"), index=False)

    with open(os.path.join(DATA_DIR, "ALL_CODES_50.txt"), "w") as f:
        f.write("\n".join(ALL_CODES_50))

    print("v2 dataset created (longer notes):")
    for fname in ["train_50.csv", "dev_50.csv", "test_50.csv", "ALL_CODES_50.txt"]:
        print(" -", os.path.join(DATA_DIR, fname))
    print("\nExample note length (words):", len(train_df.iloc[0]["text"].split()))
    print("Example labels:", train_df.iloc[0]["label"])

else:
    print("v2 dataset already exists - skipping regeneration.")
    print("Location:", DATA_DIR)

v2 dataset created (longer notes):
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v2/train_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v2/dev_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v2/test_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v2/ALL_CODES_50.txt

Example note length (words): 2849
Example labels: 285.9;530.81;272.4;401.9;428.0;250.00


### 3.1 Run One: train v2 with a short reading window (BL-1 setting)


In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v2/train_50.csv \
    --validation_file ../data/mimic3_v2/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v2/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 1 \
    --num_warmup_steps 10 \
    --output_dir ../models/mock-test-v2-bl1 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 02:31:48.922827: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Generating train split: 300 examples [00:00, 5779.24 examples/s]
Generating validation split: 60 examples [00:00, 4277.94 examples/s]
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0

### 3.2 Run Two: train v2 with a much longer reading window (BL-5 setting)

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v2/train_50.csv \
    --validation_file ../data/mimic3_v2/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v2/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 1 \
    --num_warmup_steps 10 \
    --output_dir ../models/mock-test-v2-bl5 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 02:35:48.080651: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 3.3 Second check: comparing the outcome for BL-1 vs. BL-5 on the corrected dataset

Making notes longer changed the numbers compared to v1, but the comparison between BL-1 and BL-5 is still inconclusive.

Macro and micro F1 are again identical between BL-1 and BL-5 (0.017 and 0.306), just as in v1. With only one training epoch, predicted probabilities for most codes likely still fail to cross the 0.5 decision threshold these metrics rely on, so both models end up predicting "no" for nearly all codes regardless of what each one actually saw.

AUC, which doesn't depend on a threshold, shows a difference, but in the wrong direction. BL-1 has the higher macro-AUC (0.475 vs. 0.465) and micro-AUC (0.781 vs. 0.774), despite reading less of each note. Recall@8 and F1@8 are marginally higher for BL-5, but by a small margin to plausibly be noise rather than a real effect of the longer window.

Conclusion: a longer, more varied dataset alone didn't produce a reliable BL-5 advantage. Before concluding the dataset itself is the limiting factor, the next step is to rule out training length as an explanation. A single epoch may simply not be enough for either model to develop distinguishable confidence in its predictions.

In [ ]:
import subprocess
import re
import ast
import pandas as pd

def get_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Re-runs evaluation only (num_train_epochs 0) using the already-trained
    # model saved at model_dir, instead of training again. This reproduces
    # the same metrics that were printed during training, captured directly
    # from the command's output this time instead of being retyped by hand.
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/dev_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]

    result = subprocess.run(
        command,
        cwd=f"{REPO_DIR}/src",
        capture_output=True,
        text=True,
    )

    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")

    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

bl1_v2_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v2-bl1",
    data_dir=f"{REPO_DIR}/data/mimic3_v2",
    max_length=3072,
    chunk_size=128,
)

bl5_v2_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v2-bl5",
    data_dir=f"{REPO_DIR}/data/mimic3_v2",
    max_length=5120,
    chunk_size=512,
)

comparison_df_v2 = pd.DataFrame(
    {"BL-1 (3,072 tokens)": bl1_v2_metrics, "BL-5 (5,120 tokens)": bl5_v2_metrics}
).T
comparison_df_v2 = comparison_df_v2[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
comparison_df_v2

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1 (3,072 tokens)",0.017493,0.306122,0.475206,0.781217,0.594444,0.279167,0.379915
"BL-5 (5,120 tokens)",0.017493,0.306122,0.465035,0.774232,0.597778,0.279167,0.380593


## 4. Adjusting the number of train epochs and number of warmup steps for dataset

### 4.1 Run One: train v2 with a short reading window (BL-1 setting), 3 epochs

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v2/train_50.csv \
    --validation_file ../data/mimic3_v2/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v2/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v2-bl1-e3 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 02:52:14.899537: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 4.2 Run Two: train v2 with a much longer reading window (BL-5 setting), 3 epochs

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v2/train_50.csv \
    --validation_file ../data/mimic3_v2/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v2/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v2-bl5-e3 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 02:59:37.049163: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 4.3 Third check: comparing the outcome for BL-1 vs. BL-5 on 3 epochs

Increasing training from 1 epoch to 3 did not resolve the comparison. If anything, it makes the underlying problem clearer.

Macro-F1 and micro-F1 are unchanged from the 1-epoch result and still identical between BL-1 and BL-5 (0.017 and 0.306). More training didn't shift these at all, consistent with predicted probabilities remaining stuck on the same side of the 0.5 decision threshold regardless of how long training runs. Recall@8, precision@8 and F1@8 have now also converged to be exactly identical between the two models.

The one metric that does move is macro-AUC, and it moves the wrong way: BL-1 pulls further ahead of BL-5 (0.560 vs. 0.484), a larger gap than at 1 epoch, despite BL-1 reading less of each note. If longer training were revealing a genuine advantage from the longer reading window, this gap should shrink or reverse in BL-5's favour, not widen against it.

Conclusion: more epochs cannot produce a meaningful BL-1 vs. BL-5 comparison here. Combined with the earlier random-placement result, this points to the dataset itself as the limiting factor rather than training length. The mock notes don't contain a genuine long-range signal for a longer reading window to find.

The next dataset version addresses this directly: rather than placing signal sentences at random and hoping some land past the cutoff, it deliberately assigns half the codes a guaranteed short-range position and half a guaranteed long-range position beyond 3,072 tokens, so the two reading windows are structurally different in what they can see.

In [ ]:
import subprocess
import re
import ast
import pandas as pd

def get_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Re-runs evaluation only (num_train_epochs 0) using the already-trained
    # model saved at model_dir, instead of training again. This reproduces
    # the same metrics that were printed during training, captured directly
    # from the command's output this time instead of being retyped by hand.
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/dev_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]

    result = subprocess.run(
        command,
        cwd=f"{REPO_DIR}/src",
        capture_output=True,
        text=True,
    )

    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")

    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

bl1_v2_e3_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v2-bl1-e3",
    data_dir=f"{REPO_DIR}/data/mimic3_v2",
    max_length=3072,
    chunk_size=128,
)

bl5_v2_e3_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v2-bl5-e3",
    data_dir=f"{REPO_DIR}/data/mimic3_v2",
    max_length=5120,
    chunk_size=512,
)

comparison_df_v2_e3 = pd.DataFrame(
    {"BL-1 (3,072 tokens)": bl1_v2_e3_metrics, "BL-5 (5,120 tokens)": bl5_v2_e3_metrics}
).T
comparison_df_v2_e3 = comparison_df_v2_e3[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
comparison_df_v2_e3

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1 (3,072 tokens)",0.017493,0.306122,0.559590,0.787857,0.594722,0.28125,0.381897
"BL-5 (5,120 tokens)",0.017493,0.306122,0.484461,0.790417,0.594722,0.28125,0.381897


## 5. Build adjusted synthetic dataset: real signal and deliberate short-vs-long placement

In [ ]:
import os
import random
import pandas as pd
from transformers import AutoTokenizer

DATA_DIR = f"{REPO_DIR}/data/mimic3_v3"
MODEL_NAME = "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext"

if not os.path.exists(os.path.join(DATA_DIR, "train_50.csv")):

    random.seed(42)
    tokenizer = AutoTokenizer.from_pretrained(MODEL_NAME)

    ALL_CODES_50 = [
        "401.9", "428.0", "427.31", "414.01", "584.9", "250.00", "486", "285.9",
        "38.93", "V58.61", "038.9", "996.62", "599.0", "244.9", "530.81",
        "272.4", "276.1", "285.1", "V45.81", "36.15", "99.04", "88.72",
        "V15.82", "574.20", "998.59", "39.61", "96.71", "36.06",
        "410.71", "780.09", "997.1", "424.0", "V45.82", "45.13", "584.5",
        "45.16", "276.2", "424.1", "511.9", "412", "N39.0", "78.52",
        "790.29", "780.2", "008.45", "790.6", "263.9", "V44.21", "560.1",
    ]
    ALL_CODES_50 = list(dict.fromkeys(ALL_CODES_50))[:50]

    CODE_SIGNALS = {
        "401.9":  "Blood pressure remains persistently elevated at 165/95 on repeat measurement.",
        "428.0":  "Echocardiogram demonstrates reduced ejection fraction consistent with heart failure.",
        "427.31": "Telemetry reveals an irregularly irregular rhythm consistent with atrial fibrillation.",
        "414.01": "Coronary angiogram shows significant stenosis of the left anterior descending artery.",
        "584.9":  "Creatinine has risen sharply from baseline, consistent with acute kidney injury.",
        "250.00": "Fingerstick glucose readings remain uncontrolled despite standing insulin regimen.",
        "486":    "Chest radiograph demonstrates a new consolidation consistent with pneumonia.",
        "285.9":  "Hemoglobin is low without an identified source of active bleeding.",
        "38.93":  "A central venous catheter was placed under ultrasound guidance without complication.",
        "V58.61": "Patient continues on long-term anticoagulation for prior thromboembolic event.",
        "038.9":  "Blood cultures returned positive, consistent with bloodstream infection.",
        "996.62": "Surgical site shows erythema and purulent drainage concerning for infected hardware.",
        "599.0":  "Urinalysis demonstrates pyuria and bacteriuria consistent with urinary tract infection.",
        "244.9":  "TSH is markedly elevated, consistent with underlying hypothyroidism.",
        "530.81": "Patient reports worsening reflux symptoms with regurgitation after meals.",
        "272.4":  "Lipid panel shows markedly elevated LDL cholesterol.",
        "276.1":  "Serum sodium returned low at 128, consistent with hyponatremia.",
        "285.1":  "Acute drop in hemoglobin following procedure, consistent with acute blood loss anemia.",
        "V45.81": "Patient has a prior coronary artery bypass graft noted on surgical history.",
        "36.15":  "Patient underwent single-vessel coronary artery bypass grafting during this admission.",
        "99.04":  "Patient received a packed red blood cell transfusion for symptomatic anemia.",
        "88.72":  "Transthoracic echocardiogram was performed to assess cardiac function.",
        "V15.82": "Documented history of tobacco use is noted in the social history.",
        "574.20": "Right upper quadrant ultrasound demonstrates cholelithiasis without wall thickening.",
        "998.59": "Post-operative course complicated by a wound infection requiring antibiotics.",
        "39.61":  "Patient required cardiopulmonary bypass during the operative procedure.",
        "96.71":  "Patient was placed on mechanical ventilation for respiratory failure.",
        "36.06":  "A drug-eluting stent was placed in the culprit coronary vessel.",
        "410.71": "Troponin is elevated with dynamic ST-segment changes, consistent with myocardial infarction.",
        "780.09": "Patient exhibited an altered level of consciousness on initial evaluation.",
        "997.1":  "Post-procedural course was complicated by a cardiac complication requiring monitoring.",
        "424.0":  "Echocardiogram demonstrates severe mitral regurgitation.",
        "V45.82": "History notes prior percutaneous coronary intervention with stent placement.",
        "45.13":  "Upper endoscopy was performed to evaluate the source of gastrointestinal symptoms.",
        "584.5":  "Acute kidney injury is attributed to reduced renal perfusion in the setting of illness.",
        "45.16":  "Colonoscopy was performed as part of the diagnostic workup.",
        "276.2":  "Arterial blood gas demonstrates metabolic acidosis with elevated lactate.",
        "424.1":  "Auscultation reveals a murmur consistent with aortic valve pathology.",
        "511.9":  "Chest imaging demonstrates a pleural effusion of uncertain etiology.",
        "412":    "History is notable for a prior myocardial infarction with no current acute changes.",
        "N39.0":  "Repeat urinalysis remains positive, consistent with a recurrent urinary tract infection.",
        "78.52":  "Patient required emergent cardiopulmonary resuscitation during this admission.",
        "790.29": "Point-of-care glucose testing reveals abnormal readings requiring further workup.",
        "780.2":  "Patient experienced a syncopal episode prior to admission.",
        "008.45": "Stool studies are positive for Clostridioides difficile infection.",
        "790.6":  "Basic metabolic panel reveals an electrolyte abnormality requiring correction.",
        "263.9":  "Nutritional assessment indicates evidence of malnutrition.",
        "V44.21": "Patient care requires management of a tracheostomy present from prior admission.",
        "560.1":  "Abdominal imaging is consistent with a bowel obstruction.",
    }

    FILLER_SENTENCES = [
        "Patient remains hemodynamically stable overnight without acute events.",
        "Vital signs are within normal limits at this time of assessment.",
        "Nursing notes indicate the patient tolerated the shift without incident.",
        "Physical therapy evaluated the patient and recommended continued mobility exercises.",
        "Patient reports adequate pain control on the current medication regimen.",
        "Family was updated at bedside regarding the current plan of care.",
        "Intake and output remain balanced over the past 24-hour period.",
        "Patient was seen and examined; no new complaints at this time.",
        "Diet was advanced as tolerated without difficulty.",
        "Social work is assisting with discharge planning and placement.",
        "Wound care team recommends continued daily dressing changes.",
        "Case management continues to coordinate follow-up appointments.",
        "Patient ambulated in the hallway with standby assistance today.",
        "Overnight course was unremarkable with no acute changes noted.",
        "Medication reconciliation was completed on admission.",
        "Patient and family verbalize understanding of the discharge instructions.",
        "Respiratory status remains stable on room air.",
        "Patient denies any new symptoms since the last assessment.",
        "Sleep was reportedly adequate overnight per nursing documentation.",
        "Continue current plan of care and reassess in the morning.",
    ]

    # Pre-compute exact token lengths with the real tokenizer, so placement
    # is measured in actual tokens rather than an approximated word count.
    FILLER_TOKEN_LENS = {s: len(tokenizer.encode(s, add_special_tokens=False)) for s in FILLER_SENTENCES}
    SIGNAL_TOKEN_LENS = {c: len(tokenizer.encode(s, add_special_tokens=False)) for c, s in CODE_SIGNALS.items()}

    # v3 change: rather than giving every code the same random amount of
    # padding (v2's uniform randint(30, 80)), half the codes are fixed as
    # SHORT_RANGE (their signal always lands well inside BL-1's 3,072-token
    # window) and half are fixed as LONG_RANGE (their signal is always
    # pushed past 3,072 tokens, but still well inside the 5,120-token
    # BL-5 window). This makes the two windows structurally different in
    # what they can see, rather than leaving it to chance - so BL-1 should
    # now systematically underperform on LONG_RANGE codes while BL-5 should
    # be able to pick them up.
    shuffled_codes = ALL_CODES_50.copy()
    random.shuffle(shuffled_codes)
    midpoint = len(shuffled_codes) // 2
    SHORT_RANGE_CODES = set(shuffled_codes[:midpoint])
    LONG_RANGE_CODES = set(shuffled_codes[midpoint:])

    LONG_RANGE_TARGET_MIN = 3300
    LONG_RANGE_TARGET_MAX = 3600
    TOTAL_LENGTH_CAP = 4800  # stay well clear of BL-5's 5,120 truncation point

    def add_filler(segments, current_tokens, count):
        for _ in range(count):
            s = random.choice(FILLER_SENTENCES)
            segments.append(s)
            current_tokens += FILLER_TOKEN_LENS[s]
        return current_tokens

    def add_filler_until(segments, current_tokens, target_tokens):
        while current_tokens < target_tokens:
            s = random.choice(FILLER_SENTENCES)
            segments.append(s)
            current_tokens += FILLER_TOKEN_LENS[s]
        return current_tokens

    def build_note(assigned_codes):
        segments = []
        tok_count = 0
        tok_count = add_filler(segments, tok_count, random.randint(2, 6))

        short_codes = [c for c in assigned_codes if c in SHORT_RANGE_CODES]
        long_codes = [c for c in assigned_codes if c in LONG_RANGE_CODES]
        random.shuffle(short_codes)
        random.shuffle(long_codes)

        # Short-range signals: small padding, guaranteed to sit well inside
        # the 3,072-token BL-1 window (visible to both BL-1 and BL-5).
        for code in short_codes:
            tok_count = add_filler(segments, tok_count, random.randint(2, 10))
            segments.append(CODE_SIGNALS[code])
            tok_count += SIGNAL_TOKEN_LENS[code]

        # Long-range signals: one large padding block pushes past BL-1's
        # real 3,072-token cutoff using actual tokenized length, landing in
        # a random spot within the 3,300-3,600 token zone - still well
        # inside BL-5's 5,120-token window.
        if long_codes:
            target = random.randint(LONG_RANGE_TARGET_MIN, LONG_RANGE_TARGET_MAX)
            tok_count = add_filler_until(segments, tok_count, target)
            for code in long_codes:
                tok_count = add_filler(segments, tok_count, random.randint(2, 10))
                segments.append(CODE_SIGNALS[code])
                tok_count += SIGNAL_TOKEN_LENS[code]

        # Trailing filler, capped so the note stays comfortably under 5,120.
        trailing_target = min(tok_count + random.randint(50, 200), TOTAL_LENGTH_CAP)
        tok_count = add_filler_until(segments, tok_count, trailing_target)

        return " ".join(segments)

    CODE_WEIGHTS = [1.0 / (rank + 1) for rank in range(len(ALL_CODES_50))]

    def pick_labels_for_one_note():
        target_count = random.randint(1, 6)
        picked = set()
        while len(picked) < target_count:
            picked.add(random.choices(ALL_CODES_50, weights=CODE_WEIGHTS, k=1)[0])
        return list(picked)

    def make_one_row(admission_id):
        labels = pick_labels_for_one_note()
        return {
            "SUBJECT_ID": random.randint(10000, 99999),
            "HADM_ID": admission_id,
            "text": build_note(labels),
            "label": ";".join(labels),
        }

    def make_split(number_of_notes, starting_id):
        rows = [make_one_row(starting_id + i) for i in range(number_of_notes)]
        return pd.DataFrame(rows, columns=["SUBJECT_ID", "HADM_ID", "text", "label"])

    os.makedirs(DATA_DIR, exist_ok=True)

    train_df = make_split(1500, starting_id=1)
    dev_df   = make_split(300,  starting_id=10000)
    test_df  = make_split(300,  starting_id=20000)

    train_df.to_csv(os.path.join(DATA_DIR, "train_50.csv"), index=False)
    dev_df.to_csv(os.path.join(DATA_DIR, "dev_50.csv"), index=False)
    test_df.to_csv(os.path.join(DATA_DIR, "test_50.csv"), index=False)

    with open(os.path.join(DATA_DIR, "ALL_CODES_50.txt"), "w") as f:
        f.write("\n".join(ALL_CODES_50))

    print("v3 dataset created (short-range vs. long-range signal placement):")
    for fname in ["train_50.csv", "dev_50.csv", "test_50.csv", "ALL_CODES_50.txt"]:
        print(" -", os.path.join(DATA_DIR, fname))
    print(f"\nShort-range codes ({len(SHORT_RANGE_CODES)}):", sorted(SHORT_RANGE_CODES))
    print(f"Long-range codes ({len(LONG_RANGE_CODES)}):", sorted(LONG_RANGE_CODES))

    # Sanity check: confirm actual tokenized note lengths land where intended.
    sample_note = train_df.iloc[0]["text"]
    sample_len = len(tokenizer.encode(sample_note, add_special_tokens=False))
    print(f"\nExample note actual tokenized length: {sample_len}")
    print("Example labels:", train_df.iloc[0]["label"])

else:
    print("v3 dataset already exists - skipping regeneration.")
    print("Location:", DATA_DIR)

/usr/local/lib/python3.13/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


v3 dataset created (short-range vs. long-range signal placement):
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v3/train_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v3/dev_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v3/test_50.csv
 - /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/data/mimic3_v3/ALL_CODES_50.txt

Short-range codes (24): ['008.45', '038.9', '276.1', '276.2', '36.15', '39.61', '401.9', '414.01', '424.0', '511.9', '560.1', '574.20', '584.9', '599.0', '780.09', '790.29', '88.72', '96.71', '996.62', '998.59', 'V44.21', 'V45.81', 'V45.82', 'V58.61']
Long-range codes (25): ['244.9', '250.00', '263.9', '272.4', '285.1', '285.9', '36.06', '38.93', '410.71', '412', '424.1', '427.31', '428.0', '45.13', '45.16', '486', '530.81', '584.5', '78.52', '780.2', '790.6', '99.04', '997.1', 'N39.0', 'V15.82']

Example note actual tokenized length: 3772
Example labels: 38.93;401.9;780.09;428.0;584.5


### 5.1 Run One: train v3 with a short reading window (BL-1 setting)

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl1 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 03:47:37.818316: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Generating train split: 1500 examples [00:00, 4313.39 examples/s]
Generating validation split: 300 examples [00:00, 4390.32 examples/s]
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob":

### 5.2 Run Two: train v3 with a much longer reading window (BL-5 setting)

In [ ]:
%cd {REPO_DIR}/src
!python3 run_icd.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl5 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-23 04:22:14.639382: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 5.3 Fourth check: comparing the outcome for BL-1 vs. BL-5 with real signal and deliberate short-vs-long placement

Unlike v1 and v2, macro-F1 and micro-F1 are no longer identical between the two models. BL-5 leads on both (0.020 vs. 0.015 macro-F1, 0.283 vs. 0.258 micro-F1). The first time the thresholded metrics have actually differed between BL-1 and BL-5, rather than collapsing to the same value. This is the clearest sign yet that the two models are genuinely making different predictions rather than defaulting to the same "no" answer for almost every code.

BL-5 also leads on micro-AUC, recall@8, precision@8 and F1@8, all by small but consistent margins. Macro-AUC is the one exception, where BL-1 is still marginally ahead (0.552 vs. 0.542), but that gap has narrowed considerably compared to the earlier random-placement experiments, where it widened to as much as 0.076 in BL-1's favour. BL-5 now leads on five of the six metrics.

Conclusion: deliberately guaranteeing that half the codes' evidence falls beyond BL-1's 3,072-token window produces a modest advantage for the longer reading window. Exactly the effect that random placement and additional training epochs failed to produce. The absolute scores remain low, which is expected given the small scale of this synthetic dataset (1,500 training notes, 50 codes, 3 epochs) rather than a real measure of diagnostic performance, the values here are not directly comparable to the paper's reported results on the full MIMIC-III dataset. What matters is the direction and consistency of the gap: it confirms that when a genuine long-range dependency exists in the text, a longer reading window helps recover it, supporting the paper's central claim on a controlled, synthetic benchmark.

In [ ]:
import subprocess
import re
import ast
import pandas as pd

def get_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Re-runs evaluation only (num_train_epochs 0) using the already-trained
    # model saved at model_dir, instead of training again. This reproduces
    # the same metrics that were printed during training, captured directly
    # from the command's output this time instead of being retyped by hand.
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/dev_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]

    result = subprocess.run(
        command,
        cwd=f"{REPO_DIR}/src",
        capture_output=True,
        text=True,
    )

    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")

    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

bl1_v3_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v3-bl1",
    data_dir=f"{REPO_DIR}/data/mimic3_v3",
    max_length=3072,
    chunk_size=128,
)

bl5_v3_metrics = get_metrics_from_saved_model(
    model_dir=f"{REPO_DIR}/models/mock-test-v3-bl5",
    data_dir=f"{REPO_DIR}/data/mimic3_v3",
    max_length=5120,
    chunk_size=512,
)

comparison_df_v3 = pd.DataFrame(
    {"BL-1 (3,072 tokens)": bl1_v3_metrics, "BL-5 (5,120 tokens)": bl5_v3_metrics}
).T
comparison_df_v3 = comparison_df_v3[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
comparison_df_v3

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1 (3,072 tokens)",0.015199,0.257598,0.551836,0.771774,0.558556,0.250833,0.346198
"BL-5 (5,120 tokens)",0.020408,0.282540,0.541947,0.778228,0.567056,0.253750,0.350608


## 6. Chunk-size ablation (Experiments, extra depth)

### 6.1 Chunk-size sweep on v3

In [ ]:
import os
import subprocess

def train_if_needed(output_dir, chunk_size, max_length=5120, num_train_epochs=3):
    model_file = os.path.join(output_dir, "model.safetensors")
    if os.path.exists(model_file):
        print(f"Skipping {output_dir} — model already exists.")
        return

    print(f"Training {output_dir} (chunk_size={chunk_size}) ...")
    command = [
        "python3", "run_icd.py",
        "--train_file", "../data/mimic3_v3/train_50.csv",
        "--validation_file", "../data/mimic3_v3/dev_50.csv",
        "--code_50",
        "--code_file", "../data/mimic3_v3/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
        "--per_device_train_batch_size", "1",
        "--gradient_accumulation_steps", "4",
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", str(num_train_epochs),
        "--num_warmup_steps", "20",
        "--output_dir", output_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]
    subprocess.run(command, cwd=f"{REPO_DIR}/src")

# 1024 dropped: BERT's encoder has a hard 512-token-per-chunk limit
# (max_position_embeddings = 512), so chunk_size > 512 isn't a valid
# configuration for this model and will fail immediately.
chunk_runs = {
    128: f"{REPO_DIR}/models/mock-test-v3-bl5-chunk128",
    256: f"{REPO_DIR}/models/mock-test-v3-bl5-chunk256",
    512: f"{REPO_DIR}/models/mock-test-v3-bl5",  # existing BL-5 run
}

for chunk_size, output_dir in chunk_runs.items():
    train_if_needed(output_dir, chunk_size)

Training /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/models/mock-test-v3-bl5-chunk128 (chunk_size=128) ...
Training /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/models/mock-test-v3-bl5-chunk256 (chunk_size=256) ...
Skipping /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/models/mock-test-v3-bl5 — model already exists.


### 6.2 Comparing chunk sizes at max_length 5,120 (v3 dataset)

Chunk size 128 is clearly the weakest configuration - it comes last on every single metric in the table, with no exceptions. This matches the paper's own finding that splitting a fixed-length input into many small chunks fragments local context and hurts performance compared to fewer, longer chunks.

Chunk 256 and chunk 512 are much closer to each other, and the picture between them is mixed rather than a clean "bigger is always better" story. Macro-AUC and micro-AUC keep improving as chunk size increases (0.508 → 0.533 → 0.542 macro-AUC across 128 → 256 → 512), consistent with the paper's conclusion that longer chunks help. Macro-F1 and micro-F1 are identical between 256 and 512, so those metrics don't distinguish the two configurations at all here. Recall@8, precision@8 and F1@8 actually peak slightly at chunk 256 rather than 512, though by a small enough margin (a few thousands) that it's more likely noise from a single small-scale training run than a genuine reversal.

Conclusion: the sweep reproduces the paper's headline direction - chunk size 128 measurably underperforms both larger options, confirming that a fixed total reading window still needs long enough chunks to preserve local context. Chunk 512 comes out ahead or tied on the threshold-independent metrics (AUC), which is the paper's own basis for preferring 512 over 128 as the chunk size in its best model (BL-5). The near-identical results between 256 and 512 suggest most of that benefit is already captured by 256 on this dataset, though drawing a firm conclusion there would need a larger-scale run than this synthetic, small-sample setup allows.

In [ ]:
chunk_size_configs = {
    "Chunk 128": (f"{REPO_DIR}/models/mock-test-v3-bl5-chunk128", 128),
    "Chunk 256": (f"{REPO_DIR}/models/mock-test-v3-bl5-chunk256", 256),
    "Chunk 512": (f"{REPO_DIR}/models/mock-test-v3-bl5", 512),
}

chunk_size_metrics = {}
for label, (model_dir, chunk_size) in chunk_size_configs.items():
    chunk_size_metrics[label] = get_metrics_from_saved_model(
        model_dir=model_dir,
        data_dir=f"{REPO_DIR}/data/mimic3_v3",
        max_length=5120,
        chunk_size=chunk_size,
    )

chunk_size_df = pd.DataFrame(chunk_size_metrics).T
chunk_size_df = chunk_size_df[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
chunk_size_df

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
Chunk 128,0.015199,0.257598,0.507571,0.769538,0.552444,0.248750,0.343039
Chunk 256,0.020408,0.282540,0.533283,0.777834,0.581111,0.257083,0.356466
Chunk 512,0.020408,0.282540,0.541947,0.778228,0.567056,0.253750,0.350608


## 7. Add test_50.csv evaluation support

In [ ]:
def get_test_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Same as get_metrics_from_saved_model, but evaluates on test_50.csv
    # instead of dev_50.csv. Used only once, for final reporting, never
    # for tuning decisions (that stays on the dev set).
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/test_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]

    result = subprocess.run(
        command,
        cwd=f"{REPO_DIR}/src",
        capture_output=True,
        text=True,
    )

    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")

    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

### 7.1 Final results table — key takeaways (test set)

BL-5 (5,120/512) beats BL-1 (3,072/128) on six of the seven metrics: Macro/micro-F1, micro-AUC, recall@8, precision@8 and F1@8. Macro-AUC is the one exception, and it still favours BL-1, but only by 0.003 (0.554 vs. 0.551), a much smaller gap than at earlier stages. Overall this shows the reading-window effect on held-out data.

Looking more closely at the numbers, I noticed BL-1 and BL-5-chunk128 land on exactly the same macro/micro-F1 (0.015 / 0.260), while BL-5-chunk512 and BL-5-chunk256 also match each other exactly, just at a higher value (0.020 / 0.287). Since BL-1 always uses chunk_size 128 and BL-5 uses chunk_size 512, it looks like chunk size, not window length by itself, is driving a good part of these threshold-based results. That fits what the chunk-size ablation already showed: chunk 128 is the weakest setting no matter the window length, and 256/512 perform almost the same as each other, like as on the dev set.


In [ ]:
final_configs = {
    "BL-1  (window 3,072, chunk 128)":   ("mock-test-v3-bl1", 3072, 128),
    "BL-5  (window 5,120, chunk 512)":   ("mock-test-v3-bl5", 5120, 512),
    "BL-5  (window 5,120, chunk 256)":   ("mock-test-v3-bl5-chunk256", 5120, 256),
    "BL-5  (window 5,120, chunk 128)":   ("mock-test-v3-bl5-chunk128", 5120, 128),
}

final_test_metrics = {}
for label, (model_dir_name, max_len, chunk) in final_configs.items():
    final_test_metrics[label] = get_test_metrics_from_saved_model(
        model_dir=f"{REPO_DIR}/models/{model_dir_name}",
        data_dir=f"{REPO_DIR}/data/mimic3_v3",
        max_length=max_len,
        chunk_size=chunk,
    )

final_results_df = pd.DataFrame(final_test_metrics).T
final_results_df = final_results_df[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]
final_results_df

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1 (window 3,072, chunk 128)",0.014983,0.259701,0.553821,0.782313,0.583222,0.244167,0.344224
"BL-5 (window 5,120, chunk 512)",0.020408,0.286656,0.550628,0.792222,0.585000,0.245833,0.346189
"BL-5 (window 5,120, chunk 256)",0.020408,0.286656,0.540821,0.789134,0.596500,0.249583,0.351919
"BL-5 (window 5,120, chunk 128)",0.014983,0.259701,0.502785,0.779698,0.571722,0.241667,0.339730


## 8. Bonus experiment - chunk overlap

In [7]:
#Copy of run_icd.py to perform overlap testing
!cp {REPO_DIR}/src/run_icd.py {REPO_DIR}/src/run_icd_overlap.py

### 8.1 BL-1, chunk size 128, Overlap 10%

In [10]:
%cd {REPO_DIR}/src
!python3 run_icd_overlap.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --chunk_overlap 13 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl1-overlap10 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-24 01:00:23.680913: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
Generating train split: 1500 examples [00:01, 1000.63 examples/s]
Generating validation split: 300 examples [00:01, 287.36 examples/s]
config.json: 100% 385/385 [00:00<00:00, 1.89MB/s]
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,

### 8.2 BL-1, chunk size 128, Overlap 20%

In [12]:
%cd {REPO_DIR}/src
!python3 run_icd_overlap.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 3072 \
    --chunk_size 128 \
    --chunk_overlap 26 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl1-overlap20 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-24 01:50:25.673584: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 8.3 BL-5, chunk size 512, Overlap 10%

In [13]:
%cd {REPO_DIR}/src
!python3 run_icd_overlap.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --chunk_overlap 51 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl5-overlap10 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-24 02:51:25.400740: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 8.4 BL-5, chunk size 512, Overlap 20%

In [14]:
%cd {REPO_DIR}/src
!python3 run_icd_overlap.py \
    --train_file ../data/mimic3_v3/train_50.csv \
    --validation_file ../data/mimic3_v3/dev_50.csv \
    --code_50 \
    --code_file ../data/mimic3_v3/ALL_CODES_50.txt \
    --max_length 5120 \
    --chunk_size 512 \
    --chunk_overlap 102 \
    --model_name_or_path microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext \
    --per_device_train_batch_size 1 \
    --gradient_accumulation_steps 4 \
    --per_device_eval_batch_size 1 \
    --num_train_epochs 3 \
    --num_warmup_steps 20 \
    --output_dir ../models/mock-test-v3-bl5-overlap20 \
    --model_type bert \
    --model_mode laat

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW/src
2026-09-24 03:57:05.902051: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
loading configuration file config.json from cache at /root/.cache/huggingface/hub/models--microsoft--BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext/snapshots/e1354b7a3a09615f6aba48dfad4b7a613eef7062/config.json
Model config BertConfig {
  "_name_or_path": "microsoft/BiomedNLP-PubMedBERT-base-uncased-abstract-fulltext",
  "architectures": [
    "BertForMaskedLM"
  ],
  "attention_probs_dropout_prob": 0.1,
  "classifier_dropout": null,
  "hidden_act": "gelu",
  "hidden_dropout_prob": 0.1,
  "hidden_size": 768,
  "id2label": {
    "0": "LABEL_0",
    "1": "LABEL_1",
    "2": "LABEL_2",
    "3": "LABEL_3",
    "4": "LA

### 8.5 Comparison of all the overlap testings

In [16]:
import subprocess, re, ast
import pandas as pd

def get_metrics_from_saved_model(model_dir, data_dir, max_length, chunk_size):
    # Re-runs evaluation only (num_train_epochs 0) using the already-trained
    # model saved at model_dir, instead of training again.
    command = [
        "python3", "run_icd.py",
        "--train_file", f"{data_dir}/train_50.csv",
        "--validation_file", f"{data_dir}/dev_50.csv",
        "--code_50",
        "--code_file", f"{data_dir}/ALL_CODES_50.txt",
        "--max_length", str(max_length),
        "--chunk_size", str(chunk_size),
        "--model_name_or_path", model_dir,
        "--per_device_eval_batch_size", "1",
        "--num_train_epochs", "0",
        "--output_dir", model_dir,
        "--model_type", "bert",
        "--model_mode", "laat",
    ]
    result = subprocess.run(command, cwd=f"{REPO_DIR}/src", capture_output=True, text=True)
    match = re.search(r"metrics:\s*(\{.*\})", result.stdout + result.stderr)
    if not match:
        print(result.stdout[-2000:])
        print(result.stderr[-2000:])
        raise ValueError("No metrics line found in the command output above.")
    metrics_text = re.sub(r"np\.float(?:32|64)\(([^)]+)\)", r"\1", match.group(1))
    return ast.literal_eval(metrics_text)

DATA_DIR = f"{REPO_DIR}/data/mimic3_v3"

overlap_models = {
    "BL-1, 0% overlap":  (f"{REPO_DIR}/models/mock-test-v3-bl1",           3072, 128),
    "BL-1, 10% overlap": (f"{REPO_DIR}/models/mock-test-v3-bl1-overlap10", 3072, 128),
    "BL-1, 20% overlap": (f"{REPO_DIR}/models/mock-test-v3-bl1-overlap20", 3072, 128),
    "BL-5, 0% overlap":  (f"{REPO_DIR}/models/mock-test-v3-bl5",           5120, 512),
    "BL-5, 10% overlap": (f"{REPO_DIR}/models/mock-test-v3-bl5-overlap10", 5120, 512),
    "BL-5, 20% overlap": (f"{REPO_DIR}/models/mock-test-v3-bl5-overlap20", 5120, 512),
}

overlap_metrics = {
    name: get_metrics_from_saved_model(model_dir=path, data_dir=DATA_DIR,
                                        max_length=max_len, chunk_size=chunk_sz)
    for name, (path, max_len, chunk_sz) in overlap_models.items()
}

overlap_comparison = pd.DataFrame(overlap_metrics).T
overlap_comparison[
    ["f1_macro", "f1_micro", "auc_macro", "auc_micro", "rec_at_8", "prec_at_8", "f1_at_8"]
]

,f1_macro,f1_micro,auc_macro,auc_micro,rec_at_8,prec_at_8,f1_at_8
"BL-1, 0% overlap",0.015199,0.257598,0.551836,0.771774,0.558556,0.250833,0.346198
"BL-1, 10% overlap",0.015199,0.257598,0.558446,0.771308,0.552444,0.248750,0.343039
"BL-1, 20% overlap",0.015199,0.257598,0.537901,0.771275,0.552444,0.248750,0.343039
"BL-5, 0% overlap",0.020408,0.282540,0.541947,0.778228,0.567056,0.253750,0.350608
"BL-5, 10% overlap",0.038018,0.346587,0.556694,0.790280,0.578389,0.255833,0.354752
"BL-5, 20% overlap",0.061224,0.454286,0.570052,0.800142,0.603833,0.268750,0.371954


In [11]:
!git add .
!git commit -m "Add PLM-ICD reimplementation, synthetic dataset and experiment results"
!git branch

[master b780afd] Add PLM-ICD reimplementation, synthetic dataset and experiment results
 104 files changed, 863192 insertions(+), 3 deletions(-)
 create mode 100644 HTIN Assignment 1_final.ipynb
 create mode 100644 data/mimic3_v1/ALL_CODES_50.txt
 create mode 100644 data/mimic3_v1/dev_50.csv
 create mode 100644 data/mimic3_v1/test_50.csv
 create mode 100644 data/mimic3_v1/train_50.csv
 create mode 100644 data/mimic3_v2/ALL_CODES_50.txt
 create mode 100644 data/mimic3_v2/dev_50.csv
 create mode 100644 data/mimic3_v2/test_50.csv
 create mode 100644 data/mimic3_v2/train_50.csv
 create mode 100644 data/mimic3_v3/ALL_CODES_50.txt
 create mode 100644 data/mimic3_v3/dev_50.csv
 create mode 100644 data/mimic3_v3/test_50.csv
 create mode 100644 data/mimic3_v3/train_50.csv
 create mode 100644 models/mock-test-v1-bl1/config.json
 create mode 100644 models/mock-test-v1-bl1/model.safetensors
 create mode 100644 models/mock-test-v1-bl1/special_tokens_map.json
 create mode 100644 models/mock-test-v1-

In [10]:
!git config --global user.email "jwindegger@gmail.com"
!git config --global user.name "Julia Windegger"

In [12]:
!git push -u origin master

fatal: could not read Username for 'https://github.com': No such device or address


In [ ]:
from getpass import getpass
username = "JuliaWind0"
token = getpass("Paste your GitHub Personal Access Token: ")

!git push https://{username}:{token}@github.com/JuliaWind0/HTIN5005-Assignment1_JW.git master

Paste your GitHub Personal Access Token: ··········
Enumerating objects: 95, done.
Counting objects: 100% (95/95), done.
Delta compression using up to 2 threads


In [4]:
%cd /content/drive/MyDrive/USYD_HTIN/Assignment_1_JW
!git status

/content/drive/MyDrive/USYD_HTIN/Assignment_1_JW
Refresh index: 100% (113/113), done.
On branch master
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	modified:   HTIN Assignment 1_final.ipynb

no changes added to commit (use "git add" and/or "git commit -a")
